external-plugins/rehearse/events/events.go

```diff
@@ -0,0 +1,397 @@
+package events
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"reflect"
+	"strings"
+
+	v1 "k8s.io/test-infra/prow/client/clientset/versioned/typed/prowjobs/v1"
+
+	"k8s.io/test-infra/prow/pjutil"
+
+	"github.com/sirupsen/logrus"
+	prowapi "k8s.io/test-infra/prow/apis/prowjobs/v1"
+	"k8s.io/test-infra/prow/config"
+	git2 "k8s.io/test-infra/prow/git/v2"
+	"k8s.io/test-infra/prow/github"
+)
+
+type GitHubEventsServer struct {
+	prowClient       *v1.ProwV1Client
+	logger           *logrus.Entry
+	tokenGenerator   func() []byte
+	gitClientFactory git2.ClientFactory
+	github           *github.Client
+	dryRun           bool
+	prowConfigPath   string
+	jobsConfigBase   string
+}
+
+func NewGitHubEventsServer(
+	logger *logrus.Entry,
+	prowClient *v1.ProwV1Client,
+	gitClientFactory git2.ClientFactory,
+	tokenGenerator func() []byte,
+	github *github.Client,
+	dryRun bool,
+	prowConfigPath,
+	jobsConfigBase string) *GitHubEventsServer {
+
+	return &GitHubEventsServer{
+		prowClient:       prowClient,
+		logger:           logger,
+		gitClientFactory: gitClientFactory,
+		github:           github,
+		dryRun:           dryRun,
+		tokenGenerator:   tokenGenerator,
+		prowConfigPath:   prowConfigPath,
+		jobsConfigBase:   jobsConfigBase,
+	}
+}
+
+type eventValidatingFunc func(http.ResponseWriter, *http.Request, func() []byte) (string, string, []byte, bool, int)
+
+func (s *GitHubEventsServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
```

Comment:
nit: regarding readability: I'm not a big fan of one character function arguments, as when you scroll the function declaration out of sight, you easily tend to forget what the s, w, r stood for.

Reply:
+1, I totally agree with that. Fixed.